spock-core/src/main/java/org/spockframework/runtime/extension/IStore.java

```diff
@@ -0,0 +1,289 @@
+package org.spockframework.runtime.extension;
+
+import org.spockframework.util.Beta;
+import org.spockframework.util.Checks;
+import org.spockframework.util.ReflectionUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * A hierarchical store key-value pairs.
+ * @since 2.4
+ */
+@Beta
+public interface IStore {
+  /**
+   * Get the value that is stored under the supplied {@code key}.
+   *
+   * <p>If no value is stored in the current {@link IStore}
+   * for the supplied {@code key}, ancestors of the context will be queried
+   * for a value with the same {@code key} in the {@code Namespace} used
+   * to create this store.
+   *
+   * <p>For greater type safety, consider using {@link #get(Object, Class)}
+   * instead.
+   *
+   * @param key the key; never {@code null}
+   * @return the value; potentially {@code null}
+   * @see #get(Object, Class)
+   * @see #getOrDefault(Object, Class, Object)
+   */
+  Object get(Object key);
+
+  /**
+   * Get the value of the specified required type that is stored under
+   * the supplied {@code key}.
+   *
+   * <p>If no value is stored in the current {@link IStore}
+   * for the supplied {@code key}, ancestors of the context will be queried
+   * for a value with the same {@code key} in the {@code Namespace} used
+   * to create this store.
+   *
+   * @param key          the key; never {@code null}
+   * @param requiredType the required type of the value; never {@code null}
+   * @param <V>          the value type
+   * @return the value; potentially {@code null}
+   * @see #get(Object)
+   * @see #getOrDefault(Object, Class, Object)
+   */
+  <V> V get(Object key, Class<V> requiredType);
+
+  /**
+   * Get the value of the specified required type that is stored under
+   * the supplied {@code key}, or the supplied {@code defaultValue} if no
+   * value is found for the supplied {@code key} in this store or in an
+   * ancestor.
+   *
+   * <p>If no value is stored in the current {@link IStore}
+   * for the supplied {@code key}, ancestors of the context will be queried
+   * for a value with the same {@code key} in the {@code Namespace} used
+   * to create this store.
+   *
+   * @param key          the key; never {@code null}
+   * @param requiredType the required type of the value; never {@code null}
+   * @param defaultValue the default value
+   * @param <V>          the value type
+   * @return the value; potentially {@code null}
+   * @see #get(Object, Class)
+   */
+  default <V> V getOrDefault(Object key, Class<V> requiredType, V defaultValue) {
```

Comment:
The requiredType could be a supertype of the default value.

Reply:
What would that change?
Any receiver expecting the supertype can also cope with the subtype as return type.
Maybe I just don't see it.
Can you maybe show an example where you would need `requiredType` and could not simply leave it out?

Reply:
```
ns1Root.put("key", [])
ns1Root.<String>get("key").toCharArray()
```
Fails with
```
groovy.lang.MissingMethodException: No signature of method: java.util.ArrayList.toCharArray() is applicable for argument types: () values: []
Possible solutions: toArray(), toArray(), toArray([Ljava.lang.Object;), toArray([Ljava.lang.Object;)
```
instead of
```
Caused by: org.spockframework.runtime.extension.IStore$StoreException: Object stored under key [key] is not of required type [java.lang.String]
```

Reply:
Yes, and it will be even more expressive once https://github.com/junit-team/junit5/pull/3601 gets merged.